File: Structural/Flyweight/Cpp/CharacterFactory.cpp
// CharacterFactory.cpp

#include "CharacterFactory.h"

std::unordered_map<char, FlyweightCharacter*> CharacterFactory::characterPool;

FlyweightCharacter* CharacterFactory::getCharacter(char symbol) {
  if (characterPool.find(symbol) == characterPool.end()) {
    characterPool[symbol] = new FlyweightCharacter(symbol);
  }
  return characterPool[symbol];
}

int CharacterFactory::getPoolSize() {
  return characterPool.size();
}

void CharacterFactory::cleanup() {
  for (auto& pair : characterPool) {
    delete pair.second;
  }
  characterPool.clear();
}

File: Structural/Flyweight/Python/character_factory.py
# character_factory.py

from flyweight_character import FlyweightCharacter

class CharacterFactory:
  _character_pool = {}

  @staticmethod
  def get_character(symbol: str):
    if symbol not in CharacterFactory._character_pool:
      CharacterFactory._character_pool[symbol] = FlyweightCharacter(symbol)
    return CharacterFactory._character_pool[symbol]

  @staticmethod
  def get_pool_size():
    return len(CharacterFactory._character_pool)
